gen_breaktest.go

```diff
@@ -0,0 +1,220 @@
+// +build generate
+
+// This program generates grapheme_break_test.go from the Unicode Character
+// Database auxiliary data files at https://www.unicode.org/Public/
+// Either directly via HTTP by URL or from a local copy of the file.
+//
+//go:generate go run gen_breaktest.go
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"go/format"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+)
+
+// See https://www.unicode.org/license.html for the Unicode license agreement.
+
+// We want to test against a specific version rather than the latest, which
+// can be found at:
+// https://www.unicode.org/Public/UCD/latest/ucd/auxiliary/GraphemeBreakTest.txt
+// When/if the package is upgraded to a new version, change these to generate
+// new tests.
+const (
+	url      = `https://www.unicode.org/Public/13.0.0/ucd/auxiliary/GraphemeBreakTest.txt`
+	filename = `GraphemeBreakTest-13.0.0.txt`
+)
+
+func main() {
+	log.SetPrefix("gen_breaktest: ")
+	log.SetFlags(0)
+
+	// Read text of testcases and parse into Go source code.
+	src, err := readAndParse()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Format the Go code.
+	srcfmt, err := format.Source(src)
+	if err != nil {
+		log.Fatalln("gofmt:", err)
+		//srcfmt = src
+	}
+
+	// Write it out.
+	if err := ioutil.WriteFile("grapheme_break_test.go", srcfmt, 0644); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// readAndParse reads a GraphemeBreakTest text file, either from a local file or
+// from a URL.
+//
+// It parses the file data into Go source code representing the testcases.
+func readAndParse() ([]byte, error) {
+	var r io.ReadCloser
+	if f, err := os.Open(filename); err == nil {
+		log.Printf("using %q", filename)
+		r = f
+	} else if !errors.Is(err, os.ErrNotExist) {
+		return nil, err
+	} else {
+		log.Printf("using %q", url)
+		resp, err := http.Get(url)
+		if err != nil {
+			return nil, err
+		}
+		r = resp.Body
+	}
+	defer r.Close()
+
+	buf := new(bytes.Buffer)
+	buf.Grow(120 << 10)
+	buf.WriteString(`// Code generated by go generate from gen_breaktest.go. DO NOT EDIT.
+
+package uniseg
+
+// unicodeTestCases are Grapheme testcases taken from
+// ` + url + `,
+// see https://www.unicode.org/license.html for the Unicode license agreement.
+var unicodeTestCases = []testcase {
```

Comment:
I'm getting an error in the generated test file I believe because `testcase` should be `testCase`. Could you please change this and re-generate the test file?